extension/src/emulator/server/language-server.ts

```diff
@@ -15,83 +20,141 @@ const RESTART_SERVER = 'cadence.server.flow.restart'
 const RELOAD_CONFIGURATION = 'cadence.server.flow.reloadConfiguration'
 
 export class LanguageServerAPI {
-  client!: LanguageClient
-  running: boolean
+  client: LanguageClient | null = null
+  #clientLock = new Mutex()
+  running: boolean = false
+
+  #initializedClient = false
+  #emulatorConnected = false
+
   accessCheckMode: string
   flowCommand: string
 
-  #initializedClient: boolean
-
   constructor () {
     const settings = Settings.getWorkspaceSettings()
     this.accessCheckMode = settings.accessCheckMode
     this.flowCommand = settings.flowCommand
 
-    // Init running state with false and update, when client is connected to server
-    this.running = false
-    this.#initializedClient = false
-
     void this.startClient()
+    void this.watchEmulator()
   }
 
   deactivate (): void {
-    void this.client.stop()
+    void this.client?.stop()
       .catch((err) => { void err })
   }
 
+  async watchEmulator (): Promise<void> {
+    const seconds = 3
+    setInterval(async () => {
+      const emulatorFound = await this.scanForEmulator()
+
+      if (!this.#emulatorConnected && emulatorFound) {
+        void window.showInformationMessage('Connecting to local flow emulator')
+      } else if (this.#emulatorConnected && !emulatorFound) {
+        console.log('Flow Emulator Disconnected')
+      } else {
+        return
+      }
+  
+      // Restart LS
+      void this.stopClient()
+      void this.startClient()
+    }, 1000 * seconds)
+  }
+
+  async scanForEmulator (): Promise<boolean> {
```

Comment:
The name doesn't indicate boolean result, better something like:
```suggestion
  async emulatorExists (): Promise<boolean> {
```